ext/opentelemetry-ext-flask/src/opentelemetry/ext/flask/__init__.py

```diff
@@ -24,74 +24,17 @@ def instrument_app(flask):
     """
 
     wsgi = flask.wsgi_app
-
-    def wrapped_app(environ, start_response):
-        # We want to measure the time for route matching, etc.
-        # In theory, we could start the span here and use update_name later
-        # but that API is "highly discouraged" so we better avoid it.
-        environ[_ENVIRON_STARTTIME_KEY] = time_ns()
-
-        def _start_response(status, response_headers, *args, **kwargs):
-            span = flask_request.environ.get(_ENVIRON_SPAN_KEY)
-            if span:
-                otel_wsgi.add_response_attributes(
-                    span, status, response_headers
-                )
-            else:
-                logger.warning(
-                    "Flask environ's OpenTelemetry span missing at _start_response(%s)",
-                    status,
-                )
-            return start_response(status, response_headers, *args, **kwargs)
-
-        return wsgi(environ, _start_response)
-
-    flask.wsgi_app = wrapped_app
-
+    flask.wsgi_app = otel_wsgi.OpenTelemetryMiddleware(wsgi)
     flask.before_request(_before_flask_request)
-    flask.teardown_request(_teardown_flask_request)
 
 
 def _before_flask_request():
-    environ = flask_request.environ
-    span_name = flask_request.endpoint or otel_wsgi.get_default_span_name(
-        environ
-    )
-    parent_span = propagators.extract(
-        otel_wsgi.get_header_from_environ, environ
+    wsgi_tracer = trace.tracer_source().get_tracer(
+        otel_wsgi.__name__, otel_wsgi.__version__
     )
-
-    tracer = trace.tracer_source().get_tracer(__name__, __version__)
-
-    attributes = otel_wsgi.collect_request_attributes(environ)
+    span = wsgi_tracer.get_current_span()
+    if flask_request.endpoint:
+        span.update_name(flask_request.endpoint)
```

Comment:
The whole point of not using the WSGI middleware was to avoid using `update_name` which is discouraged and indeed did not work at the time with the backend I used. For example, you can't configure sampling based on the endpoint, if you use `update_name`.

Reply:
Got it, that's good info. Sorry I missed that comment.

When you say it's use is discouraged, where is that stated? The specification states:

```
The Span's start and end timestamps reflect the elapsed real time of the operation. A Span's start time SHOULD be set to the current time on span creation. After the Span is created, it SHOULD be possible to change the its name, set its Attributes, and add Links and Events. These MUST NOT be changed after the Span's end time has been set.
```

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md

Looking at the comment it's stated that this make it hard for dynatrace's span grouping. Have alternatives been tried that may be compatible with this, e.g. grouping on the exporter side?

Reply:
Here is the wording in the spec: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/api-tracing.md#updatename

> It is highly discouraged to update the name of a Span after its creation. Span name is often used to group, filter and identify the logical groups of spans. And often, filtering logic will be implemented before the Span creation for performance reasons. Thus the name update may interfere with this logic.

For Dynatrace, this is no longer as important as it was at the time, but what the spec says still holds.

Reply:
got it. I'd argue against exposing update_name in the first place then, since the existence of that API means that people will probably use it, and would then require a processor to handle that case anyway.

I'm going to take a stab at seeing if I can start the span in the flask_before_request, and end it with the wsgi, which should avoid the use of update_name.